kratos/spatial_containers/search_wrapper.cpp

```diff
@@ -0,0 +1,334 @@
+        } else {
+            noalias(local_max) = ZeroVector(3);
+            noalias(local_min) = ZeroVector(3);
+        }
+
+        // Assign local BB
+        for (int i = 0; i < 3; ++i) {
+            local_bounding_box[2 * i] = local_max[i];
+            local_bounding_box[2 * i + 1] = local_min[i];
+        }
+
+        // Gather all bounding boxes
+        mrDataCommunicator.AllGather(local_bounding_box, mGlobalBoundingBoxes);
+    }
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+template<class TSearchObject>
+std::vector<int> SearchWrapper<TSearchObject>::RansksPointIsInsideBoundingBox(const array_1d<double, 3>& rCoords)
+{
+    std::vector<int> ranks;
+    const int world_size = GetWorldSize();
+    std::array<double, 6> local_bb;
+    const auto it_begin = mGlobalBoundingBoxes.begin();
+    for (int i = 0; i < world_size; ++i) {
+        auto vec_it = it_begin + 6 * i;
+        for (unsigned int j = 0; j < 6; ++j, ++vec_it) {
+            local_bb[j] = *vec_it;
+        }
+        if (SearchUtilities::PointIsInsideBoundingBox(local_bb, rCoords)) {
+            ranks.push_back(i);
+        }
+    }
+
+    return ranks;
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+template<class TSearchObject>
+std::vector<int> SearchWrapper<TSearchObject>::RansksPointIsInsideBoundingBoxWithTolerance(
+    const array_1d<double, 3>& rCoords,
+    const double Tolerance
+    )
+{
+    std::vector<int> ranks;
+    const int world_size = GetWorldSize();
+    std::array<double, 6> local_bb;
+    std::vector<double> bb_tolerance(mGlobalBoundingBoxes.size());
+    SearchUtilities::ComputeBoundingBoxesWithToleranceCheckingNullBB(mGlobalBoundingBoxes, Tolerance, bb_tolerance);
+    const auto it_begin = bb_tolerance.begin();
+    for (int i = 0; i < world_size; ++i) {
+        auto vec_it = it_begin + 6 * i;
+        for (unsigned int j = 0; j < 6; ++j, ++vec_it) {
+            local_bb[j] = *vec_it;
+        }
+        if (SearchUtilities::PointIsInsideBoundingBox(local_bb, rCoords)) {
+            ranks.push_back(i);
+        }
+    }
+
+    return ranks;
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+template<class TSearchObject>
+void SearchWrapper<TSearchObject>::LocalSearchInRadius(
+    const PointType& rPoint,
+    const double Radius,
+    std::vector<ResultType>& rResults,
+    const int AllocationSize
+    )
+{
+    // Get the rank
+    const int rank = GetRank();
+
+    // If we are using GeometricalObjectBins we can use the optimized search
+    if constexpr (IsGeometricalObjectBins) {
```

Comment:
Wouldn't be easier for this to just have a specialized version? They don't share any code

Reply:
yes, they sahre part of the code, I am not using that constexpr all  the time

Reply:
i see

```
auto rank = getRank()
if constexpr(x) {
   do_something()
} else {
   do_another_thing()
}
```
What is being shared?